### PR TITLE
(VDB-1144) Enable diffs to come in out of order

### PR DESCRIPTION
**Pull request by Gslaughl**
This PR enables the auction (flip/flap/flop) trigger tables to react appropriately to diffs coming in out of order. Reorg-safety still has not been implemented.

All of flop/flap's triggers are just copy-pasted flip triggers with some columns removed, plus a `:%s/flip/flop/g` (though flip has some triggers that the other two don't). So it might not be necessary for reviewers to go through all 3 migrations with a fine-tooth comb, since they're all almost identical.

### CODE REVIEW

File: transformers/test_data/shared_behaviors/shared_flip_bid_trigger_suite.go [hunk 1]
```diff
@@ -0,0 +1,407 @@
...
+	Expect(actualBid.Bid).To(Equal(expectedBid.Bid))
+	Expect(actualBid.Updated).To(Equal(expectedBid.Updated))
+}
+
+func assertFlipSnapshot(actualFlip, expectedFlip flipBidSnapshot, expectedBlockNumber int64) {
+	assertBidSnapshot(actualFlip.commonBidSnapshot, expectedFlip.commonBidSnapshot, expectedBlockNumber)
+	Expect(actualFlip.Usr).To(Equal(expectedFlip.Usr))
+	Expect(actualFlip.Gal).To(Equal(expectedFlip.Gal))
+	Expect(actualFlip.Tab).To(Equal(expectedFlip.Tab))
+}
```
Comment (Gslaughl) [inline transformers/test_data/shared_behaviors/shared_flip_bid_trigger_suite.go]: We already have datatypes/functions similar to these in `test_helpers`, but these are slightly different since they're designed to work with the `maker.flip`-style records from the DB, whereas the ones in `test_helpers` work with the `api.flip_state`-style types (returned by `api.get_flip`, etc.).

So I think it makes sense to have 2 distinct sets of helpers for 2 distinct types. Although, hopefully we'll eventually be able to expose the `flip` table directly instead of exposing its data via `api.get_flips` and `all_flips`, at which point we'd be able to delete the other set of helpers.

File: transformers/test_data/shared_behaviors/shared_flip_bid_trigger_suite.go [hunk 2]
```diff
@@ -0,0 +1,407 @@
...
+				queryErr := db.Select(&valueHistory, getFieldQuery)
+				Expect(queryErr).NotTo(HaveOccurred())
+				Expect(len(valueHistory)).To(Equal(2))
+				Expect(valueHistory[1].String).To(Equal(initialColumnVal))
+			})
+		})
+	})
+}
+
+func FlipBidSnapshotTriggerTests(input BidTriggerTestInput) {
```
Comment (Gslaughl) [inline transformers/test_data/shared_behaviors/shared_flip_bid_trigger_suite.go]: I had some trouble deciding how to balance abstraction vs readability with these tests. I ended up allowing some duplication for these 2 tests (the `inserting a new field` suite is duplicated on line 236) since they have to make assertions about the entire record, and thus are dealing with 2 distinct types for `flip` vs `flap/flop` records. Whereas the `updating history with a new value` suite only deals with 1 field at a time.

So, while it would've been _possible_ to abstract the suites on lines 163 and 263 into a single suite, I felt like it would be too detrimental to the tests' (already questionable) clarity.

File: transformers/test_data/shared_behaviors/shared_flip_bid_trigger_suite.go [hunk 3]
```diff
@@ -0,0 +1,407 @@
...
+			initialBidValues[input.Metadata.Name] = updatedRepoVal
+			expectedBid := commonBidSnapshotFromValues(bidID, headerTwo.BlockNumber, addressID, headerOne.Timestamp,
+				initialBidValues)
+			assertBidSnapshot(bidSnapshots[0], expectedBid, headerOne.BlockNumber)
+			assertSingleField(bidSnapshots[0], expectedBid, input.ColumnName)
+		})
+	})
+}
+
+func randomBidStorageValue(valueType vdbStorage.ValueType, packedValueType vdbStorage.ValueType) (interface{}, string) {
```
Comment (Gslaughl) [inline transformers/test_data/shared_behaviors/shared_flip_bid_trigger_suite.go]: This function was my best effort at keeping packed-value repository logic out of the body of the tests. It returns 2 values: the first is what the `repo.Create` function wants to accept, and the second is the string value that will actually be inserted into the trigger-table, which we can use in assertions or raw SQL inserts. Usually these 2 values will be the same, except in the case of packed values.

File: transformers/test_data/shared_behaviors/shared_flip_bid_trigger_suite.go [hunk 4]
```diff
@@ -0,0 +1,407 @@
...
+			var parseErr error
+			bidID, parseErr = strconv.Atoi(input.Metadata.Keys[constants.BidId])
+			Expect(parseErr).NotTo(HaveOccurred())
+			var addressErr error
+			addressID, addressErr = shared.GetOrCreateAddress(input.ContractAddress, db)
+			Expect(addressErr).NotTo(HaveOccurred())
+		})
+
+		Describe("updating history with new value", func() {
+			It("updates field in subsequent blocks", func() {
```
Comment (Gslaughl) [inline transformers/test_data/shared_behaviors/shared_flip_bid_trigger_suite.go]: This whole suite follows the general pattern of:
- insert record 1
- insert record 2
- see how record 1 reacted

File: transformers/test_data/shared_behaviors/shared_flip_bid_trigger_suite.go (hunk 3 above)
Comment (rmulhol) [reply transformers/test_data/shared_behaviors/shared_flip_bid_trigger_suite.go]: Might be worth exploring how much clutter we'd create by extracting a separate suite/function for packed values? Totally on board with the de-duplicating approach, but wondering if this is a place where there's a need for some kind of polymorphism to handle important differences between the data structures/avoid returning vals that are only useful to a subset of callers

File: transformers/test_data/shared_behaviors/shared_flip_bid_trigger_suite.go (hunk 1 above)
Comment (Gslaughl) [reply transformers/test_data/shared_behaviors/shared_flip_bid_trigger_suite.go]: Actually no, it doesn't look like we do. For some reason I thought I added stories like that at the same time that I added the ones about making better use of the ilk + urn trigger tables, but I guess not 🤔. I'll see to getting some into the backlog.

File: transformers/test_data/shared_behaviors/shared_flip_bid_trigger_suite.go [hunk 5]
```diff
@@ -0,0 +1,407 @@
...
+				Expect(err).NotTo(HaveOccurred())
+
+				var valueHistory []sql.NullString
+				queryErr := db.Select(&valueHistory, getFieldQuery)
+				Expect(queryErr).NotTo(HaveOccurred())
+				Expect(len(valueHistory)).To(Equal(2))
+				Expect(valueHistory[1].String).To(Equal(newColumnVal))
+			})
+
+			It("ignores rows from blocks after the next time the field is updated", func() {
```
[the remark this answers is not in the document]
Comment (Gslaughl) [reply transformers/test_data/shared_behaviors/shared_flip_bid_trigger_suite.go]: Using `repo.Create` inserts a record into the storage table as well as into the trigger-table, and the trigger needs the storage table record to figure out the max `block_number` of the records it should be updating.

File: transformers/storage/flap/repository_test.go [hunk 6]
```diff
@@ -319,12 +319,22 @@ var _ = Describe("Flap storage repository", func() {
...
+			triggerInput := shared_behaviors.BidTriggerTestInput{
+				Repository:      repository,
+				Metadata:        bidBidMetadata,
+				ContractAddress: repository.ContractAddress,
+				TriggerTable:    constants.FlapTable,
+				FieldTable:      constants.FlapBidBidTable,
+				ColumnName:      constants.BidColumn,
+			}
+			shared_behaviors.CommonBidSnapshotTriggerTests(triggerInput)
+			shared_behaviors.SharedBidHistoryTriggerTests(triggerInput)
```
[the remark this answers is not in the document]
Comment (Gslaughl) [reply transformers/storage/flap/repository_test.go]: Yeah although it isn't very clear. Since there are 2 separate bid structures at play, i.e. the one used by flap + flop vs the one used by flip (where the flap/flop one is a subset of the flip one), I'm trying to use the dichotomy between a "common bid" vs a "flip bid" to express that distinction. Hence why we're using `CommonBidSnapshot...` in this suite and `FlipBidSnapshot...` in the flip storage suite.

OTOH, by `SharedBidHistoryTriggerTests`, I just mean "tests about behavior that's shared across all bid history triggers".

So to recap, `Common` is being used as a way to distinguish between types of bids, and `Shared` is being used to describe what the tests are about. I'll try to think of a less confusing way to imply those ideas haha. Open to suggestions!

File: transformers/storage/flap/repository_test.go [hunk 7]
```diff
@@ -25,14 +25,13 @@ var _ = Describe("Flap storage repository", func() {
 	var (
 		db                   = test_config.NewTestDB(test_config.NewTestNode())
 		flapContractAddress  = test_data.FlapAddress()
-		repository           flap.FlapStorageRepository
+		repository           = &flap.FlapStorageRepository{ContractAddress: flapContractAddress}
```
[the remark this answers is not in the document]
Comment (Gslaughl) [reply transformers/storage/flap/repository_test.go]: Weirdly, I tried that and ended up getting a nil pointer dereference error. So now I'm afraid to touch it 😅

File: transformers/test_data/shared_behaviors/shared_flip_bid_trigger_suite.go (hunk 3 above)
Comment (Gslaughl) [reply transformers/test_data/shared_behaviors/shared_flip_bid_trigger_suite.go]: Yeah that's a good call. I guess in this case I was so preoccupied with whether or not I _could_ combine all the tests suites into one, that I didn’t stop to think if I _should_. ;)

Comment (Gslaughl) [reply transformers/test_data/shared_behaviors/shared_flip_bid_trigger_suite.go]: The other thing I considered in order to avoid this problem entirely, is to create a utility function that handles inserting storage records, rather than relying on the repository. My thinking is that the trigger tests have nothing to do with the repositories (especially the 'packed values' portion of the repo logic), so the fact that we even need to think about packed values in the trigger tests feels like sort of a code smell to me. Maybe it would be worth decoupling these tests from the repository.

File: db/migrations/20190711091453_create_flip_table.sql [hunk 8]
```diff
@@ -135,256 +166,657 @@ COMMENT ON FUNCTION flip_bid_time_created
...
+          AND flip_bid_guy.address_id = new_diff.address_id
+          AND block_number > diff_block_number);
+BEGIN
+    UPDATE maker.flip
+    SET guy = new_diff.guy
+    WHERE flip.bid_id = new_diff.bid_id
+      AND flip.address_id = new_diff.address_id
+      AND flip.block_number >= diff_block_number
+      AND (next_guy_diff_block IS NULL
+        OR flip.block_number < next_guy_diff_block);
```
[the remark this answers is not in the document]
Comment (Gslaughl) [reply db/migrations/20190711091453_create_flip_table.sql]: I did some reading up on composite indexes, and it looks like in this case Postgres will use the primary key index, since the columns we're filtering on are included in the index. Definitely a good idea to keep indexes in mind moving forward though!

File: db/migrations/20190711091453_create_flip_table.sql [hunk 9]
```diff
@@ -22,102 +23,132 @@ CREATE INDEX flip_address_index
...
     LANGUAGE sql;
 
-COMMENT ON FUNCTION get_latest_flip_bid_bid
+COMMENT ON FUNCTION flip_bid_bid_before_block
     IS E'@omit';
 
-CREATE FUNCTION get_latest_flip_bid_gal(bid_id numeric) RETURNS TEXT AS
+CREATE FUNCTION flip_bid_usr_before_block(bid_id NUMERIC, address_id INTEGER, header_id INTEGER) RETURNS TEXT AS
+$$
+SELECT usr
```
[the remark this answers is not in the document]
Comment (Gslaughl) [reply db/migrations/20190711091453_create_flip_table.sql]: Nice catch, I probably should've explained why I added this! The short answer is no, this isn't a new field. 

The `usr` field isn't in the spec, and tbh I added it on a whim because it was very easy and I wanted the flip table's structure to match [the `Bid` object](https://github.com/makerdao/dss/blob/master/src/flip.sol#L52) in the solidity code. The only real benefit I see to keeping it, is that the `usr` is an urn identifier, so having it directly available in this table will make the `urn` computed column more performant, since it won't have to do a separate lookup to find the `usr` associated with a bid.

File: db/migrations/20190711091453_create_flip_table.sql [hunk 10]
```diff
@@ -135,256 +166,657 @@ COMMENT ON FUNCTION flip_bid_time_created
...
+ON CONFLICT (block_number, bid_id, address_id) DO UPDATE SET guy = new_diff.guy;
 $$
-BEGIN
-    WITH diff_block AS (
-        SELECT block_number, block_timestamp
+    LANGUAGE sql;
+-- +goose StatementEnd
+
+-- +goose StatementBegin
+CREATE OR REPLACE FUNCTION maker.update_flip_guys_until_next_diff(new_diff maker.flip_bid_guy) RETURNS VOID
```
[the remark this answers is not in the document]
Comment (Gslaughl) [reply db/migrations/20190711091453_create_flip_table.sql]: The flip table is basically a collection of snapshots, each representing the state of a bid at a given block. So when a diff from the past is inserted, we need to update all the snapshots which would've been affected by the diff that was inserted. This trigger's job is to figure out which snapshots would be affected, and update them accordingly.

Using `flip_bid_guy` as an example: the only snapshots we care about updating are the ones from blocks in between the `flip_bid_guy` from the past (which was just inserted), and the next-most-recent `flip_bid_guy` diff. That's because the next one will essentially overwrite the `guy` field for that bid, but we still want a record of what the bid's state was before that later diff happened.

Hope that makes sense! Lmk if I can clarify anything.

[the remark this answers is not in the document]
Comment (elizabethengelman) [reply db/migrations/20190711091453_create_flip_table.sql]: 🙏 thank you, that makes sense.

File: transformers/storage/flap/repository_test.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (elizabethengelman) [reply transformers/storage/flap/repository_test.go]: I wonder if removing the `Common` and `Shared` would be okay? So just have `BidSnapshotTriggerTests` and `BidHistoryTriggerTests` and we can infer that they're common or shared by the fact that they're used in multiple places. And then `FlipBidSnapshot...` would stay named that way to explicitly point out that Flips are different?

File: transformers/test_data/shared_behaviors/shared_flip_bid_trigger_suite.go [hunk 11]
```diff
@@ -0,0 +1,407 @@
...
+			Expect(parseErr).NotTo(HaveOccurred())
+			var addressErr error
+			addressID, addressErr = shared.GetOrCreateAddress(input.ContractAddress, db)
+			Expect(addressErr).NotTo(HaveOccurred())
+		})
+
+		Describe("updating history with new value", func() {
+			It("updates field in subsequent blocks", func() {
+				_, initialColumnVal := randomBidStorageValue(input.Metadata.Type, input.PackedValueType)
+				_, setupErr := db.Exec(insertFieldQuery, addressID, headerTwo.BlockNumber, bidID, initialColumnVal, timestampTwo)
```
[the remark this answers is not in the document]
Comment (elizabethengelman) [reply transformers/test_data/shared_behaviors/shared_flip_bid_trigger_suite.go]: Not sure, but do you think it would help to clarify if the test description included something like "updates field in subsequent blocks **when no associated diff records**"? That also may make things more complicated... I sometimes get too verbose with my test descriptions. 😊

File: transformers/storage/flap/repository_test.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (Gslaughl) [reply transformers/storage/flap/repository_test.go]: Great idea! Will update. FYI I've been enacting feedback from this PR onto the reorg PR, just because a lot of the suggested changes are also applicable to that one, and it's easier to make them all at once.
